### PR TITLE
gh-114863: What's new in Python 3.13: JIT compiler

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
I mainly kept most of Guido's original text, and split them into sections. I also added more exposition in certain parts (traces, optimization, execution engine). Let me know what y'all think.

PR as a follow up to https://github.com/python/cpython/pull/114826 for discussion.

----
📚 Documentation preview 📚: https://cpython-previews--114862.org.readthedocs.build/

* Issue: gh-114863

### CODE REVIEW

File: Doc/whatsnew/3.13.rst
Comment (danielhollas) [inline Doc/whatsnew/3.13.rst]: Maybe to clarify

```suggestion
The second is machine code. When the ``--enable-experimental-jit`` build-time
```

File: Doc/whatsnew/3.13.rst
Comment (danielhollas) [inline Doc/whatsnew/3.13.rst]: Add a link to the paper, and perhaps other more accessible resources (blog posts, Brandt's talk...)?

File: Doc/whatsnew/3.13.rst
Comment (danielhollas) [inline Doc/whatsnew/3.13.rst:538]: I wonder if the guard concept could be expanded a bit more. The example of "type checking" makes things fairly clear what it means, but wonder if it could be expanded.

File: Doc/whatsnew/3.13.rst
Comment (danielhollas) [inline Doc/whatsnew/3.13.rst]: General comment: I wonder if for people who are compiler noobs but like to learn more it would be nice to link to wikipedia or other accessible sources? In this case, it is easy to google and find this: https://en.wikipedia.org/wiki/Constant_folding but in other cases it might not be?

(to be clear, I love the current text, it is very clearly written :heart:, so this does not seem as important)

**Review by danielhollas — COMMENTED** [review]
Hi from a random stranger who knows very little about compilers :wave: 

I love the write up, it is very clear! Just some suggestions / questions.

File: Doc/whatsnew/3.13.rst
Comment (Fidget-Spinner) [reply Doc/whatsnew/3.13.rst]: Thanks for the kind words! I think most of the terms are explained (I try to give a short summary of what they mean in each bullet point). Not all of this can be found in Wikipedia articles though, and some are found in particular programming language research papers.

**Comment by gvanrossum** [comment]
Looks like the right level of detail, will review later.

File: Doc/whatsnew/3.13.rst
Comment (eendebakpt) [inline Doc/whatsnew/3.13.rst]:
```suggestion
fast compilation, reported as orders of magnitudes faster versus
```
(or: small compilation times)

File: Doc/whatsnew/3.13.rst
Comment (eendebakpt) [inline Doc/whatsnew/3.13.rst]:
```suggestion
The second is the JIT. When the ``--enable-experimental-jit``
```
("machine code" is not an execution engine)

File: Doc/whatsnew/3.13.rst
Comment (hugovk) [inline Doc/whatsnew/3.13.rst]: A suggestion, but just fine to leave it as is too:
```suggestion
  optimizations, up till CPython 3.13's beta release.
```

File: Doc/whatsnew/3.13.rst
Comment (hugovk) [inline Doc/whatsnew/3.13.rst]: Similarly:
```suggestion
The final performance results will be updated before CPython 3.13's beta release.
```

**Review by hugovk — APPROVED** [review]
Looks good!

File: Doc/whatsnew/3.13.rst
Comment (mdboom) [inline Doc/whatsnew/3.13.rst]:
```suggestion
  optimizations, until CPython 3.13's beta release.
```

File: Doc/whatsnew/3.13.rst
Comment (mdboom) [inline Doc/whatsnew/3.13.rst]:
```suggestion
There are two execution engines for Tier 2 IR: the Tier 2 interpreter and the Just-in-Time (JIT) compiler.
```

It's jarring to see the JIT mentioned in the paragraph below before it is introduced, so I think introducing them both here helps.

File: Doc/whatsnew/3.13.rst
Comment (mdboom) [inline Doc/whatsnew/3.13.rst]:
```suggestion
The Tier 2 interpreter is mostly intended for debugging
```

File: Doc/whatsnew/3.13.rst
Comment (mdboom) [inline Doc/whatsnew/3.13.rst:573]:
```suggestion
The other execution engine is the JIT compiler. When the ``--enable-experimental-jit``
```

File: Doc/whatsnew/3.13.rst
Comment (mdboom) [inline Doc/whatsnew/3.13.rst]:
```suggestion
build-time dependency on `LLVM <https://llvm.org>`_. The main benefit of this technique is
```

File: Doc/whatsnew/3.13.rst
Comment (gvanrossum) [inline Doc/whatsnew/3.13.rst]: The example does not show expressions but statements.
```suggestion
  statements like ::
```

File: Doc/whatsnew/3.13.rst
Comment (gvanrossum) [inline Doc/whatsnew/3.13.rst]: Could use explanation of "forward data-flow analysis"? The second sentence is highly speculative, I would express a more short-term and concrete benefit.

File: Doc/whatsnew/3.13.rst
Comment (gvanrossum) [inline Doc/whatsnew/3.13.rst]: This last sentence reads more like an academic report about your work than as an explanation.

File: Doc/whatsnew/3.13.rst
Comment (gvanrossum) [inline Doc/whatsnew/3.13.rst:552]: IIUC this work is independent from the data-flow analysis part. It relies on dictionary watchers, which deserve a separate mention if they aren't already mentioned.

File: Doc/whatsnew/3.13.rst
Comment (gvanrossum) [inline Doc/whatsnew/3.13.rst]: I think we have to be a bit more careful here giving credit where due.

File: Doc/whatsnew/3.13.rst
Comment (gvanrossum) [inline Doc/whatsnew/3.13.rst]: Maybe this should be at the top.

File: Doc/whatsnew/3.13.rst
Comment (gvanrossum) [inline Doc/whatsnew/3.13.rst:583]: Perhaps add an explanation why fast compilation is important.

File: Doc/whatsnew/3.13.rst
Comment (gvanrossum) [inline Doc/whatsnew/3.13.rst]: I don't see the point of this section -- contributors are already listed above. If and when Microsoft wants its name in the credits I will let you know, but not now.

File: Doc/whatsnew/3.13.rst
Comment (gvanrossum) [inline Doc/whatsnew/3.13.rst]: Add "here"?

File: Doc/whatsnew/3.13.rst
Comment (gvanrossum) [inline Doc/whatsnew/3.13.rst]: Not sure we need to drag it down that much.

File: Doc/whatsnew/3.13.rst
Comment (gvanrossum) [inline Doc/whatsnew/3.13.rst]: Please name the author of the talk. (Aren't there also other info resources about the JIT compiler by @brandtbucher?)

**Review by gvanrossum — COMMENTED** [review]
TBH, I'm not convinced we should add this yet. In case we do decide to add it, here are some suggestions. But I'd like to focus the "What's new" document on useful information for Python users, plus the traditional credits ("contributed by XXX"). More details about how it works ought to go either in a .md document in the source code or maybe in the devguide (if it's valuable for maintainers) or in a personal or team blog (if it's more about academic results).

File: Doc/whatsnew/3.13.rst
Comment (Fidget-Spinner) [reply Doc/whatsnew/3.13.rst]: We had a similar credits section in 3.11, so this is mirroring that, but I will remove it. https://docs.python.org/3/whatsnew/3.11.html#about

File: Doc/whatsnew/3.13.rst
Comment (Fidget-Spinner) [reply Doc/whatsnew/3.13.rst:552]: dictionary watchers were implemented in 3.12, not 3.13. So it's already mentioned there.

File: Doc/whatsnew/3.13.rst
Comment (Fidget-Spinner) [reply Doc/whatsnew/3.13.rst]: I will remove Jules and Martin but I will insist on crediting Manuel, I've met with him consistently every 2 weeks to discuss this work (and adjacent experiments on optimizing CPython), and its technical details. This has been ongoing for over 5 months now. So I'm very thankful for his help.

File: Doc/whatsnew/3.13.rst
Comment (gvanrossum) [reply Doc/whatsnew/3.13.rst]: > We had a similar credits section in 3.11, so this is mirroring that, but I will remove it. https://docs.python.org/3/whatsnew/3.11.html#about

Thanks. It's tradition to credit individuals regardless of who paid for their time.

**Comment by Yhg1s** [comment]
Do you want to get this in before 3.13.0a4? (Scheduled for today, Tuesday, but delayed until tomorrow for the libexpat upgrade and some other fixes.)

**Comment by Fidget-Spinner** [comment]
This shouldn't hold up the release. I'm okay with getting it in anytime before 3.13 beta. It might need major revision anyways to let the user know why they benefit. What do you think @gvanrossum ?

**Comment by gvanrossum** [comment]
Let’s hold off until we can agree on it. There’s a minimal thing in there already that’s not wrong.

**Review by CAM-Gerlach — COMMENTED** [review]
As someone with modest familiarity with the topic I found it relatively straightforward to understand, and didn't find it _that_ excessive purely as far as length-wise. On the other hand, I did find it read to me bit more to me like a (well-written) short blog post, or perhaps an Explanation in Diataxis terms, rather than a What's New.

IMO, this had less to do with the high-level content, and more in terms of structure, tone and how it presents information—it seemed to put more emphasis on discussing the hows and whys of the implementation and the optimization process as opposed to focusing on the main takeaways and user-facing impacts (optimizations, new CLI flags, future changes to expect, etc).

I think a comparison to the [3.11 What's New's Faster CPython section](https://docs.python.org/3.12/whatsnew/3.11.html#faster-cpython), which was referenced in this one and which this work is in part built upon, is instructive here. That was many times longer in total length, but is much more focused (in terms of the headings, text and structure) on the user-visible differences, performance impacts and high-level changes as opposed to an explaining of the non-user-facing low-level implementation details.

Some specific observations and suggestions:

* **Observation**:The explanation of the Tier 2 IR seems to be rather long and prominent for something that's low-level internals, and focused on the implementation rather than on the main high-level changes.

  **Suggestion**: Cut it down it down to just a sentence or two in the introduction, focusing on just the basic high-level changes the ultimately impacts users—there's a new tier 2 intermediate representation for hot code that allows more complex optimizations across function call boundaries.

* **Observation**: I think it is valuable to point out the optimizations that this currently includes (like was done for 3.11), but with the focus on the impact of the optimization (i.e. how/why is this faster) rather than how it is currently written (focusing on the mechanics of how each code transformation is performed. For example, consider the very first bullet:

  > * Type propagation -- through forward   data-flow analysis <https://clang.llvm.org/docs/DataFlowAnalysisIntro.html>_, we infer and deduce information about types.

  This tells me what the mechanical transformation does internally, but skips over the much more important bit I care about as a user: How does this actually make my code faster (and what code does it affect)? Many of the others are similar in this regard.
  
  **Suggestion**: Rewrite the bullets to focus on how each optimization actually makes mode faster (and when available, what code and by how much), and keep the explanations of the mechanics brief, high-level and frame in terms of "how does this make code faster" rather than being the main focus

* **Observation**: Likewise, much of the information in the Execution Engine section is potentially useful, but the focus seems to be more on explaining the execution engines and copy and patch, with mentioning the user-impactful changes being secondary to that.

  **Suggestion**: Consider reframing this section to enumerate the specific changes in 3.13. with brief explanations being secondary, in support of those rather than being the main focus.

* **Observation** The results and future work reads, quite frankly, like what would find in a conference paper, and seems rather overblown for what it is.

  **Suggestion**: It could be replaced with a brief note at the top of the overall section that performance improvement metrics will be added around the time of the first beta, and that the JIT is experimental and optimized and lays the ground work for the future.

**Pull request closed without merging**
